pkg/controller/networkpolicy/clustergroup.go

```diff
@@ -178,34 +199,38 @@ func (n *NetworkPolicyController) syncInternalGroup(key string) error {
 		return nil
 	}
 	grp := grpObj.(*antreatypes.Group)
-	if grp.IPBlock == nil {
+	// Retrieve the ClusterGroup corresponding to this key.
+	cg, err := n.cgLister.Get(grp.Name)
+	if err != nil {
+		klog.Infof("Didn't find the ClusterGroup %s, skip processing of internal group", grp.Name)
+		return nil
+	}
+	n.processServiceReference(grp)
+	newMemberSet := controlplane.GroupMemberSet{}
+	if grp.Selector != nil {
 		// Find all Pods matching its selectors and update store.
-		groupSelector := grp.Selector
-		pods, _ := n.processSelector(groupSelector)
-		memberSet := controlplane.GroupMemberSet{}
+		pods, _ := n.processSelector(*grp.Selector)
 		for _, pod := range pods {
 			if len(pod.Status.PodIPs) == 0 {
 				// No need to insert Pod IPAddress when it is unset.
 				continue
 			}
-			memberSet.Insert(podToGroupMember(pod, true))
+			newMemberSet.Insert(podToGroupMember(pod, true))
 		}
+	}
+	// for ipBlock Groups, newMemberSet and grp.GroupMembers will both be empty.
+	if !newMemberSet.Equal(grp.GroupMembers) {
```

Comment:
@Dyanngg is the following scenario possible:
- ServiceA with selectorX selects PodA
- ClusterGroupA references ServiceA
- I update ServiceA's selector to selectorY, which also selects PodA
- Because the group members are the same, this function does not update the `Selector` field
- I now create a new PodB, selected by selectorY (but not by selectorX)
- The internal group is not updated properly to include PodB because of the incorrect selector

I may not be reading the code correctly though. Just trying to understand whether `!newMemberSet.Equal(grp.GroupMembers)` is sufficient.

Reply:
It is totally a possible scenario and a great reminder for not try to be 'smart' in writing these conditions based on some false assumptions..... Updated the logic, thanks @antoninbas for spotting this.